### ISSUE TITLE
Cassandra index super slow

### ISSUE THREAD

**Issue by luxifr** [pyr/cyanite#156]
I gave the cassandra index a try and it seems super slow... it takes about a second for a path query, which is way too high for any meaningful use...

how do I debug this?

**Comment by TheElectricCapybara** [comment]
Confirming issues with Cassandra index performance, it appears to have big issues resolving wildcards as well.

**Comment by genericgithubuser** [comment]
The new cassandra index is showing a second to few second delay here as well.  We are trying to get this spun up to support a fairly decent sized production instance, so this poses a bit of a problem since we can not go live like this.  Is the long term plan to continue to support elasticsearch as an index option?  If so, we will revert to the previous version we were testing against and continue to use elastic if the expectation is that this will be a viable configuration in future builds.  If the plan for the stable release is to do away with the elasticsearch option, then we won't go the revert route since it'd be a temporary option.  
For anyone getting good cassandra index performance, were there specific tweaks you had to make to get this to play nicely.  Just testing this with data from 1,500 hosts with an average of 30 datapoints per against a 3 node cassandra cluster is showing the performance hit, so any tricks are appreciated.  Yes, much of it will be implementation specific, but any tweaks others have had to make for this would still be helpful to be aware of.

**Comment by eckardt** [comment]
We also found that the reads from the cassandra index are slow. Our testing setup is a single server with HDD, storing ~30000 metric paths.

There is one query for getting paths matching a given pattern which is taking especially long: https://github.com/pyr/cyanite/blob/5cdda22246f4725d9fc1e44e504efcf305c70b1c/src/io/cyanite/index/cassandra.clj#L30

One tweak that helped us improve the read performance for this particular query is to enable row caching in Cassandra: Set `row_cache_size_in_mb` in `cassandra.yaml` to something suitable (the default is `0` which disables row caching) and change the caching configuration for the `metric.path` table: `ALTER TABLE metric.path WITH caching = {'keys': 'ALL', 'rows_per_partition': '10'};`

Another tweak we tried is to cache the results of these queries in-memory for a couple of seconds. I pushed the necessary changes here: https://github.com/eckardt/cyanite/tree/memoize-matches-in-index
This is rather quick and dirty, but makes the cassandra index usable for us.

**Comment by rsvancara** [comment]
Any resolution on this?  Really a show stopper for us.

**Comment by jippi** [comment]
tried Cassandra TWCS ? ( https://github.com/jeffjirsa/twcs )

**Comment by rsvancara** [comment]
How will this help slow searches.  Does it provide an indexed search??

**Comment by aww-yiss** [comment]
Part of the challenge I've experienced here seems to be that under prod load with a volume of high cardinally metrics (~15 million unique metrics/min), it seems the ability to manage the index get pushed into old gen of the heap, and then the jvm spends a lot of CPU time of its time trying to keep up with the indexing. 

Some details for reference, hopefully some useful:

cyanite.yaml
```yaml
index:
  type: cassandra
  options:
    cluster: ["node1fqdn", "node2fqdn"]
    keyspace: metric
logging:
  level: info
  console: false
  files:
    - '/var/log/cyanite/cyanite.log'
input:
  - type: carbon
    host: '0.0.0.0'
    port: 2013
store:
  cluster: ["node1fqdn", "node2fqdn"]
    keyspace: 'metric'
api:
  host: '10.0.0.10'
  port: 8080
queues:
  defaults:
    ingestq:
      pool-size: 50
      queue-capacity: 5000000
    writeq:
      pool-size: 50
      queue-capacity: 5000000
engine:
  rules:
    default: [ "1m:90d" ]
```


Oracle Java 1.8.60 options
```
java -server -ea -Xms8G -Xmx8G -Xloggc:/var/log/cyanite/gc.log -XX:+AlwaysPreTouch -XX:+HeapDumpOnOutOfMemoryError -XX:+OptimizeStringConcat -XX:+PrintGCApplicationStoppedTime -XX:+PrintGCDateStamps -XX:+PrintGCDetails -XX:+PrintHeapAtGC -XX:+PrintPromotionFailure –XX:+PrintTenuringDistribution -XX:+UseCompressedOops -XX:+UseFastAccessorMethods -XX:+UseConcMarkSweepGC -XX:+CMSScavengeBeforeRemark -XX:CMSMaxAbortablePrecleanTime=60000 -XX:CMSWaitDuration=30000 -XX:+UseGCLogFileRotation -XX:GCLogFileSize=10M -XX:NumberOfGCLogFiles=10 -XX:PrintFLSStatistics=1 -Dcom.sun.management.jmxremote -Dcom.sun.management.jmxremote.authenticate=false -Dcom.sun.management.jmxremote.local.only=false -Dcom.sun.management.jmxremote.port=9010 -Dcom.sun.management.jmxremote.ssl=false -jar /opt/cyanite/cyanite-0.5.1-standalone.jar -f /etc/cyanite.yaml
```

![trace](https://cloud.githubusercontent.com/assets/3112865/14184466/4c358734-f743-11e5-8adc-d21b5005092b.png)
![heap](https://cloud.githubusercontent.com/assets/3112865/14184467/4c41ea06-f743-11e5-8469-f67801fb2cac.png)


Have also tried tweaking jvm to use G1GC and adjusting heap sized up and down, as small as 1GB and up to 32GB, in all instances the same issue  is observed

**Comment by ifesdjeen** [comment]
@aww-yiss it's very hard to say anything without knowing which exactly version you're running. If it's not master, I'm afraid it makes a little sense to even debug it, since the problem is most likely gone (at least judging the symptoms). 

@jippi i'm afraid that TWCS won't help for this particular issue. We recommend the date tiered compaction from cassandra itself though. This issue is more related to the index tho.

@rsvancara we've implemented the brand-new SASI-based cassandra index that unfortunately requires 3.4 version of Cassandra. It will land the master shortly, we're currently testing out the branch.

**Comment by rsvancara** [comment]
@ifesdjeen We have stood up cassandra 3.4.  I have three cassandra instances in AWS ready to go and I can quickly flip the switch and send about 40K metrics per second to these instances.  What is the best way to get your changes and test them?  Do you have an updated schema file I can load into Cassandra? I have the installation all automated with Ansible playbooks so I can iterate through changes quickly.  If I can get this working, then I wont have to go down the road of looking at KairosDB, which I am already having to hack the graphite plugin code to facilitate storage of awkwardly inconsistent metric names.

**Comment by ifesdjeen** [comment]
@rsvancara let me test a couple more things today and I'll let you know the status. Don't want to spend your time before I make sure it's ready for alpha users. I'll also consult @pyr, as he'll review my patch :)

**Comment by aww-yiss** [comment]
@ifesdjeen This was from master, but a while back now. It was listed as 0.5.1 but there seems to be little consistency in the release tags, so you're probably correct in asserting that this isn't debuggable at this point. Just trying to shed some insight.

**Comment by ifesdjeen** [comment]
Thanks for doing that! There are no real release tags, all releases up till now were internal, there was no official or binary release, therefore there're no tags.

**Comment by pauzed** [comment]
Is there any way I can be a part of this alpha group?  I've been using Cyanite since 0.1.3, finally got internal buy-in to drop our old metric cluster, and move to 0.5.1, but having horrendous performance issues, even building from master as of yesterday, which includes the SASI changes.  Wildcards certainly appear to cause a significant amount of stress, worse than that is that stress becomes blocking of any other queries to cassandra. I will try out @aww-yiss 's tweaks to the cyanite conf + jvm as well.  I really enjoyed 0.1.3 but hate its inability to do proper rollups and have been looking forward to these updates for more than a year now.

**Comment by ifesdjeen** [comment]
@pauzed Sure, you can always try it out. It'd be good to get a rough description of what your metrics look like to understand where exactly the bottleneck is. 

Do wildcards cause performance degradation while building a dashboard? Could you roughly describe so I could reproduce locally?

**Comment by pauzed** [comment]
@ifesdjeen Happily.  Right now, I'm currently just piping in system metrics, from Diamond, into Cyanite, using the base Diamond collectors (cpu, network, disk space, disk usage, load average, memory, vmstat, tcp, and filestat). I'm currently only ingesting about 100 servers.  In most circumstances, I don't have much trouble using /metrics/find?query=servers.*.*.*.* for discovering metric paths, the real problem I have is rendering json content from a query like: /render?target=servers.serverA.cpu.*.user&format=json&from=epochTimeStamp  This call appears to first figure out the wild card, so grab servers.serverA.cpu.*, then make a bunch of queries for servers.serverA.cpu.cpu0.user, servers.serverA.cpu.cpu1.user, etc.  Making this render call almost always hits my 60s request timeout on Nginx. During these calls, I notice a pretty significant jump in CPU utilization on the cyanite instance I'm requesting from, as well as the 3 cassandra nodes.

This problem impacts less my ability to create a dashboard, but significantly hinders my ability to -render- the dashboard.  A single query taking 10-60s makes it nearly impossible to reliably view/visualize datapoints, and I'm the only current consumer of this 0.5.1, the only other activities happening on this cluster is ingesting the Diamond data.

I'm currently operating a 3 node Cassandra 3.5 cluster on AWS.  It's reasonably over-provisioned using m4.4xlarge instances, giving me 3x 16core 64gb ram cluster.  I'm using the Cassandra schema located under doc/schema.cql, the only change I've made was using NetworkTopologyStrategy over SimpleStrategy:

CREATE KEYSPACE metric WITH replication = {'class': 'NetworkTopologyStrategy', 'us-east': '1'}  AND durable_writes = true;

Testing this in a single node in Vagrant, I don't notice this problem, I've not tried reproducing this in a 3 node vagrant setup.  I'm currently making the GC changes to test out any behaviour change.  Currently using Oracle JDK 1.8.0_45

Hope this helps, and feel free to ping for more info, happy to provide it!  Thanks again for the great work, on a good day, I'm really happy with Cyanite :)

**Comment by ifesdjeen** [comment]
Don't worry, we'll get it to perform :) Of course first approximation turned out to be less optimal than I initially anticipated. I'll take a closer look at it over the weekend.

**Comment by MichaelHierweck** [comment]
In our test setup we currently have about 20.000 (static) path names, submitting about 1500 metrics/sec from Riemann.

Cyanite's internal index:
rather fast, lower cpu consumption, requires minimum 2 GB heap (otherwise GC takes all cpu resulting in OOM after 30-60 minutes)

Cassandra index:
really slow (unusable?), high cpu consumtion (both cyanite and cassandra), 512 MB heap seems to be enough

**Comment by ifesdjeen** [comment]
Cassandra index is the only reasonable way to work for now, there's no better alternative. I've already created a test suite to reproduce, just didn't get enough uninterrupted time to work on it in all details. Hopefully Pfingstferien allow.

**Comment by ifesdjeen** [comment]
@MichaelHierweck I've tested the Cassandra index for performance again.

I have 25K static names: 

```
cqlsh:cyanite_test> select count(*) from segment;

 count
-------
 26762
(1 rows)

cqlsh:cyanite_test> select count(*) from path;

 count
-------
 25006
```

I run a simple query that would return around 50 results (not a few, but not too many either): 

```clj
(time (index/leaves index "0.1.*"))
;; => Elapsed time: 8.895672 msecs
```

So SASI index or indexing mechanism per se is definitely not a source of a problem.

**Comment by ifesdjeen** [comment]
@pauzed I've checked for your case as well:

> This call appears to first figure out the wild card, so grab servers.serverA.cpu.\*, then make a bunch of queries for servers.serverA.cpu.cpu0.user, servers.serverA.cpu.cpu1.user, etc

This is not true, although it would've been be an easy thing to fix. We're doing query multiplication and then we perform C* query, although with `IN`, so no separate queries are done.

For now, I can not identify a source of a problem, even after running stresses with loads similar to described ones locally. If you could run the queries against your C* cluster to confirm that Cyanite is the performance bottleneck and in this case send some profiling information, it would be highly appreciated.

**Comment by MichaelHierweck** [comment]
It seems we are doing something terribly wrong as we experience both bad performance/high cpu usage and still some kind of memory leak. :(

Would you please provide your cyanite and cassandra settings? This might help us investigate...I already provided our settings. We played around with heap sizes last weak but larger heap sizes only postpone the GC/OOM trouble.

**Comment by ifesdjeen** [comment]
TBH I'd first check Cassandra... I have super-standard settings, I didn't configure pretty much anything :/ 

Is it possible to maybe find you on IRC one of these days, for ~30 min and chat / try to work through problems together? It'd also help to exclude the factors such as wrong build etc.

**Comment by MichaelHierweck** [comment]
Thank you. Is there a cyanite IRC channel?

**Comment by ifesdjeen** [comment]
Ah, no there's no channel. My handle is `ifesdjeen` though. I'm there most of time when online :)

**Comment by pauzed** [comment]
@ifesdjeen im happy to try out the queries if you would like to post them here, and how I can leverage the test pull requests to see if I can further pinpoint the performance problem I've been experiencing.

@MichaelHierweck there was a #cyanite on Freenode, but its terribly unactive, this is where I've connected with pyr in the past.

I can make availability any time to troubleshoot this as well.  Thanks again for your help.

**Comment by ifesdjeen** [comment]
I'm trying to create a universal benchmark, where one could separate read path from write path. If the problem is in combination, it's kind of ok (e.g. we would just have to figure out sharding for writes and load-balance reads).

**Comment by ifesdjeen** [comment]
@pauzed @MichaelHierweck i _might_ have just solved the mystery. Turns out that index was being blocked by constant writes into the `segment` and `path` tables, as the throttling mechanism relied on the method in `nbhm` that was supposed to query for the existence of the metric name and then write it to the database. Although because of weird API of `nbhm`, their `contains` by default had a meaning of `containsValue`. So indexing was occurring more than one and these two problems were shadowing each other (wrong placement of indexing and nbhm pretending it checks for the key)

Since my tests were running on the Cyanite instance that was idle (as I was testing features in isolation), it all didn't reveal itself.

**Comment by ifesdjeen** [comment]
Anyways, https://github.com/pyr/cyanite/pull/207/commits/66783bd0b8177904af56da2f708fb7b09c8fa03c and https://github.com/pyr/cyanite/pull/207/commits/77c6ce3bce0c5e1ff111a147eb8e25020269c512 should solve the problem.

**Comment by MichaelHierweck** [comment]
On 21.05.2016 23:52, αλεx π wrote:
> Anyways,
> https://github.com/pyr/cyanite/pull/207/commits/66783bd0b8177904af56da2f708fb7b09c8fa03c
> and
> https://github.com/pyr/cyanite/pull/207/commits/77c6ce3bce0c5e1ff111a147eb8e25020269c512
> should solve the problem.

Thanks. I'm going to try your PRs...

**Comment by ifesdjeen** [comment]
I'm closing that one for now after some measurements and some positive feedback.

**Comment by pauzed** [comment]
Hey, so sorry for not getting back on this sooner.  With the same number of instances writing metrics into Cyanite, I no longer see the same read performance hits as expressed previously.  Looking like I'm getting similar performance with Cyanite 0.1.3 that I've migrated out of.  Thanks for your attention to this matter!

**Comment by ifesdjeen** [comment]
👍 🎆 @pauzed I can't be happier! Thank you for reporting. Very good to hear that.

**Comment by MichaelHierweck** [comment]
Thanks a lot. Great work.

**Comment by pyr** [comment]
Awesome news, thanks!
